### PR TITLE
KEYCLOAK-18565 JPA roles no-downtime store

### PR DESCRIPTION
**Pull request by vramik**

### CODE REVIEW

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java
Comment (ahus1) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java:173]: I suppose this should set the searchable field name on the entity as well by adding

```
this.name = name;
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java
Comment (ahus1) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java:259]: While the setters store their data in metadata as well, I don't see this happening for the attribute field (all setter/remove instances).
For example for `removeAttribute` I would have expected a call to `metadata.removeAttribute(name);` 
Is this missing here, or is this triggered somewhere else?

**Review by ahus1 — CHANGES_REQUESTED** [review]
Hi @vramik, thank for preparing this pull request. 

I reviewed the code changes, still without running them as I don't have an environment setup of them yet. 

If you feel any comment is off-topic here, please spin it off to a separate discussion.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java:89]: This cast is dangerous. `value[0]` is expected to be `String` in line 84, and generally can be any `Collection` or `Stream`

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java:134]: Could these methods be defined in `JpaModelCriteriaBuilder`?
Perhaps if the class was updated like this:
```java
public class JpaModelCriteriaBuilder<Self> {
    private static final Function<BiFunction<CriteriaBuilder, Root<JpaRoleEntity>, Predicate>, Self> instantiator;

   // ... e.g.
    @Override
    public Self not(Self builder) {
        return instantiator.apply(
            (cb, root) -> cb.not(builder.getPredicateFunc().apply(cb, root))
        );
    }
```

and this `instantiator` could be initialized in the constructor, called from the descendants like this:

```
    JpaRoleModelCriteriaBuilder() {
        super(JpaRoleModelCriteriaBuilder::new);
    }
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/delegate/JpaRoleDelegateProvider.java
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/delegate/JpaRoleDelegateProvider.java:39]: Good catch.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java:259]: Attributes are stored as [association](https://github.com/vramik/keycloak/blob/KEYCLOAK-18565-roles/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java#L67-L68), therefore the field is omitted from json when [serialization](https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/jsonb/JsonbType.java#L78) is done (the same applies to id field).

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java:173]: This field is marked as "non-insertable" and "non-updatable": https://github.com/vramik/keycloak/blob/KEYCLOAK-18565-roles/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java#L63

It is generated from json and it's happening on database level, so when update happens it's updated only in metadata object and then database generates value of this field automatically. It's used when we want to load just subset of fields using `read(QueryParameters<RoleModel> queryParameters)`: https://github.com/vramik/keycloak/blob/KEYCLOAK-18565-roles/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleMapKeycloakTransaction.java#L77-L81

File: model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml
Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml:43]: Could you please elaborate what do you mean by _isolation level for the JPA provider_? I'm not sure if I understand correctly.

Some indexes could be marked as unique and also some unique constraints could be added. I'll investigate possible advantages or disadvantages when using those.

note: `MapRoleProvider` does not know anything about specific implementation of the store below.

File: model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java:76]: The id is `String` in terms of `MapRoleProvider`: https://github.com/vramik/keycloak/blob/KEYCLOAK-18565-roles/model/map/src/main/java/org/keycloak/models/map/common/AbstractEntity.java#L25

+1 for updating wording for exceptions.

**Comment by vramik** [comment]
Thank you very much @ahus1 for the review. I'll incorporate some of your suggestions. Please see inline for the details.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java:259]: Thanks for the explanation, I can now understand. Still it was a "surprise" for me, maybe that's something that would help future developers if it would be part of a general rule for this store (unless a field is being migrated to a searchable field in a future version, and will be in metadata so that the previous entity version finds it there as well when reading the data).

Consider this comment to be resolved.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java:173]: Thanks, this explains it for me. As this was "surprising" to me, I'd be happy to see 2-3 comments in the code what's happening there / why the annotations have been placed here and what is happening in the background, as it might help future developers/maintainers of the code. Just not too much comments in too many places to risk that they run out of sync.

File: model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java
Comment (ahus1) [reply model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java:76]: Right, the ID is a String here, I was making assumption here, sorry for that. 

Still: what would the code check for that the ID doesn't already exist - what is the scenario? That the map provider generated an ID that already exists?

File: model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml
Comment (ahus1) [reply model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml:43]: To elaborate on the isolation level: The Map storage doesn't know about the isolation level, still the results will vary depending on the isolation configured for the storage. The following assumes that all actions are performed within a single transaction.

The code in `addRealmRole` calls `getRealmRole`, and then throws an exception if the role already exists. It then later adds the role to the store (in this case the JPA store). 

Here the varying results based on the JPA store's configuration (by reading the code, not by testing, let's hope I get it right here):

Isolation level SERAILIZABLE: The database will ensure that the search result for the realm role is stable (guaranteed not to change) within the transaction. Therefore no-one can insert another role to the database with the same name and realm (the items used in the query). Therefore the `addRealmRole` can take as long as it wants to add the new role to the database and commit the change, no-one can interfere. Anyone who tries to interfere will either be blocked or will receive an exception when trying to commit.

Any other isolation level (both REPEATABLE_READ and READ_COMMITTED): After the check that the client role is not in the database at that moment, any other parallel thread can add the role with the same name to the database and commit the change. The data would then be inconsistent. REPEATABLE_READ will only ensure that rows read from the database are unchanged (no rows have been read when looking for the existing role), and READ_COMMITTED will will only ensure that any rows read previously have been committed by their previous transactions and won't be rolled back. 

One way to prevent this inconsistency in REPEATABLE_READ or READ_COMMITTED would be a database constraint or unique index. Contrary to my previous comment, the index or constraint would need to be on realm and role, not on role alone. 

On the other hand: By having this `getRealmRole` in `addRealmRole`, the map storage provider is making an implicit assumption on the storage provider that no-one else will add a role between the check and the insert. Such implicit assumptions make it hard for a provider to do the right thing here. @hmlnarik

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java:178]: I've added a javadoc to both `JpaRoleEntity` and `JpaClientEntity` where I've tried to describe the behavior.

File: model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java:76]: Yes, the check it there to verify that there is no object with that id already. AFAIK it originally comes from unique constrains of legacy jpa store. @hmlnarik please correct me if I am wrong.

File: model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml
Comment (ahus1) [reply model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml:43]: I talked to Hynek yesterday about this, and the differences we will have across different providers. 

The current code in MapRoleProvider does a best-effort to check if the role exists.

A provider implementation should see what additional checks it can implement to prevent inconsistent data. For the JPA provider this would be adding constraints and unique keys. 

As this would need to be added for both clients and roles, I would be fine to add constraints in a separate issue and pull request after this PR has been merged.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java:178]: Looks good to me, thanks!

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java:92]: The operand may be an array even when `value.length == 1`. For correct treatment, please see e.g. https://github.com/keycloak/keycloak/blob/4c747047cef126a182b78cb1238824b447c1fbc9/model/map/src/main/java/org/keycloak/models/map/storage/chm/CriteriaOperator.java#L144-L157

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java:110]: Could you just use `Objects.toString(id, null)` rather than throwing an exception?

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java]: Thank you

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java]: Thank you for the suggestion. I've played a little bit with it and came up with 2 possible solutions (https://github.com/keycloak/keycloak/pull/9415/commits/28dacf606aa1d5ad699d844f050c7d0979e706aa or https://github.com/keycloak/keycloak/pull/9415/commits/80d13940a7cfbabc295b1762800ab08169e0b8dd). 

The first approach uses `public abstract class JpaModelCriteriaBuilder<E, Self extends JpaModelCriteriaBuilder>` and child classes then `implements ModelCriteriaBuilder`. 
Methods `and`, `or` and `not` don't require casting `Self builder` into `JpaModelCriteriaBuilder` using this approach. But we cannot mark the methods by `@Override` annotation (and IDE doesn't find then as implementations, at least in my case :) )

While he second approach uses `public abstract class JpaModelCriteriaBuilder<E, M, Self extends ModelCriteriaBuilder<M, Self>> implements ModelCriteriaBuilder<M, Self>` and child classes then just extends it. 

I wonder which approach would be better, wdyt? There might be some other even better solution though.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java]: Thank you, I haven't realized that. I've updated it according to this.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java]: https://github.com/keycloak/keycloak/commit/80d13940a7cfbabc295b1762800ab08169e0b8dd is the correct one.

Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java]: Thank you

File: model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml
Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml:42]: It's used every time when exact match for the role.name is queried. It is not possible to use both gin and b-tree index for single sql statement, therefore index gin on `fRealmId` cannot be used.

The query plan then looks like:
```
Index Scan using role_name on role  (cost=0.43..8.46 rows=1 width=175) (actual time=0.037..0.038 rows=1 loops=1)
  Index Cond: ((name)::text = '94f269adc0'::text)
  Filter: ((metadata -> 'fRealmId'::text) = '"test0"'::jsonb)
Planning Time: 0.217 ms
Execution Time: 0.049 ms
```

Comment (hmlnarik) [reply model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml:42]: This looks like a good suggestion: I believe we would need it to check of uniqueness of the role name per realm / client anyway, and that can only be done if there is unique index that includes realm ID.

File: model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml
Comment (ahus1) [reply model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml:43]: As agreed with @vramik yesterday I created a new issue for adding constraints to clients: https://github.com/keycloak/keycloak/issues/9610

Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml:43]: Thank you @ahus1

File: model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml
Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml:50]: `realmId`, `clientId` and `description` generated fields was added and index was updated.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java]: I had to switch it back to throwing CriterionNotSupportedException as there is `UUID` type expected. Please let me know if it requires some update.

**Comment by vramik** [comment]
This PR requires https://github.com/keycloak/keycloak/pull/9588/ to be merged first. 

Also it doesn't contain concurrent access handling, currently discussed in https://github.com/keycloak/keycloak/discussions/9601

**Review by ahus1 — DISMISSED** [review]
All questions I asked were addressed. Given the PR "congestion" that is coming where PRs depend on each other, I'd rather like this PR to be merged than not merged.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java:106]:
```suggestion
                                in.value(UUIDKey.INSTANCE.fromString(Objects.toString(id, null)));
                            } catch (IllegalArgumentException e) {
                                throw new CriterionNotSupportedException(modelField, op, id + " is not in uuid format.", e);
```

**Review by hmlnarik — APPROVED** [review]
Thank you for the PR!

**Review by sguilhen — APPROVED** [review]
LGTM, thanks @vramik . Also tested with CockroachDB.

**Pull request merged**
